### PR TITLE
[#5657] Include error message when checking for null using Objects.requireNonNull()

### PR DESCRIPTION
**Pull request by alexbaeza**
Adds an error message whenever using `Objects.requireNonNull()` to indicate which param is not expected to be null, provides sufficient information to a user to fix the issue.

Before:
```
Caused by: java.lang.NullPointerException: 
	at java.util.Objects.requireNonNull(Objects.java:228)
```
After:
```
Caused by: java.lang.NullPointerException: 'connectionString' cannot be null.
	at java.util.Objects.requireNonNull(Objects.java:228)
```

As discussed in this issue: https://github.com/Azure/azure-sdk-for-java/issues/5657

### CODE REVIEW

File: sdk/core/azure-core/src/main/java/com/azure/core/util/Context.java [hunk 1]
```diff
@@ -40,7 +40,7 @@
      */
     public Context(Object key, Object value) {
         if (key == null) {
```
Comment (conniey) [inline sdk/core/azure-core/src/main/java/com/azure/core/util/Context.java]: I would replace Objects.requireNotNull to fit the java SDK guidelines.

File: sdk/eventgrid/microsoft-azure-eventgrid/src/main/java/com/microsoft/azure/eventgrid/customization/EventGridSubscriber.java [hunk 2]
```diff
@@ -55,7 +55,7 @@ public void putCustomEventMapping(final String eventType, final Type eventDataTy
             throw new IllegalArgumentException("eventType parameter is required and cannot be null or empty");
         }
         if (eventDataType == null) {
```
Comment (conniey) [inline sdk/eventgrid/microsoft-azure-eventgrid/src/main/java/com/microsoft/azure/eventgrid/customization/EventGridSubscriber.java]: Would change this to Objects.requireNotnull

File: sdk/eventhubs/microsoft-azure-eventhubs-eph/src/main/java/com/microsoft/azure/eventprocessorhost/EventProcessorHost.java [hunk 3]
```diff
@@ -381,173 +381,173 @@ public void uncaughtException(Thread t, Throwable e) {
          * all instances consuming from the same Event Hub and consumer group. The name must be unique because
          * it is used to distinguish which instance owns the lease for a given partition of the event hub. An
          * easy way to generate a unique host name is to call EventProcessorHost.createHostName("mystring").
-         * 
+         *
          * @param hostName  a name for this host instance. See method notes.
-         * @param consumerGroupName  the consumer group on the Event Hub 
+         * @param consumerGroupName  the consumer group on the Event Hub
```
Comment (conniey) [inline sdk/eventhubs/microsoft-azure-eventhubs-eph/src/main/java/com/microsoft/azure/eventprocessorhost/EventProcessorHost.java]:
```suggestion
         * @param consumerGroupName  The consumer group for the Event Hub.
```

File: sdk/eventhubs/microsoft-azure-eventhubs-eph/src/main/java/com/microsoft/azure/eventprocessorhost/EventProcessorHost.java [hunk 4]
```diff
@@ -381,173 +381,173 @@ public void uncaughtException(Thread t, Throwable e) {
...
         }
-        
+
         public interface ManagerStep {
             /**
              * Use the built-in Azure Storage-based lease and checkpoint managers.
-             * 
-             * @param storageConnectionString  connection string for an Azure Storage account  
+             *
+             * @param storageConnectionString  connection string for an Azure Storage account
```
Comment (conniey) [inline sdk/eventhubs/microsoft-azure-eventhubs-eph/src/main/java/com/microsoft/azure/eventprocessorhost/EventProcessorHost.java]:
```suggestion
             * @param storageConnectionString  Connection string for the Azure Storage account.
```

File: sdk/eventhubs/microsoft-azure-eventhubs-eph/src/main/java/com/microsoft/azure/eventprocessorhost/EventProcessorHost.java [hunk 5]
```diff
@@ -381,173 +381,173 @@ public void uncaughtException(Thread t, Throwable e) {
...
              */
             AuthStep useAzureStorageCheckpointLeaseManager(String storageConnectionString, String storageContainerName, String storageBlobPrefix);
-            
+
             /**
              * Use the built-in Azure Storage-based lease and checkpoint managers.
-             * 
-             * @param storageCredentials         credentials for an Azure Storage account, such as an AAD token  
+             *
+             * @param storageCredentials         credentials for an Azure Storage account, such as an AAD token
```
Comment (conniey) [inline sdk/eventhubs/microsoft-azure-eventhubs-eph/src/main/java/com/microsoft/azure/eventprocessorhost/EventProcessorHost.java]: These shouldn't be aligned but that's for another PR..

File: sdk/eventhubs/microsoft-azure-eventhubs/src/main/java/com/microsoft/azure/eventhubs/RetryPolicy.java [hunk 6]
```diff
@@ -26,7 +26,7 @@ protected RetryPolicy(final String name) {
 
     public static boolean isRetryableException(Exception exception) {
         if (exception == null) {
-            throw new IllegalArgumentException("exception cannot be null");
+            throw new IllegalArgumentException("exception cannot be null.");
```
Comment (conniey) [inline sdk/eventhubs/microsoft-azure-eventhubs/src/main/java/com/microsoft/azure/eventhubs/RetryPolicy.java]: Replace with Objects.requireNotNull.

File: sdk/eventhubs/microsoft-azure-eventhubs/src/main/java/com/microsoft/azure/eventhubs/impl/EventDataBatchImpl.java [hunk 7]
```diff
@@ -37,7 +37,7 @@ public int getSize() {
     public boolean tryAdd(final EventData eventData) throws PayloadSizeExceededException {
 
         if (eventData == null) {
-            throw new IllegalArgumentException("eventData cannot be null");
+            throw new IllegalArgumentException("eventData cannot be null.");
```
Comment (conniey) [inline sdk/eventhubs/microsoft-azure-eventhubs/src/main/java/com/microsoft/azure/eventhubs/impl/EventDataBatchImpl.java]: Replace with Objects.requireNotNull.

File: sdk/eventhubs/microsoft-azure-eventhubs/src/main/java/com/microsoft/azure/eventhubs/impl/EventDataImpl.java [hunk 8]
```diff
@@ -114,7 +114,7 @@ public EventDataImpl(byte[] data) {
         this();
 
         if (data == null) {
-            throw new IllegalArgumentException("data cannot be null");
+            throw new IllegalArgumentException("data cannot be null.");
```
Comment (conniey) [inline sdk/eventhubs/microsoft-azure-eventhubs/src/main/java/com/microsoft/azure/eventhubs/impl/EventDataImpl.java]: Replace with Objects.requireNotNull.

File: sdk/servicebus/microsoft-azure-servicebus/src/main/java/com/microsoft/azure/servicebus/management/EntityNameHelper.java [hunk 9]
```diff
@@ -82,7 +82,7 @@ static void checkValidRuleName(String ruleName) {
 
     private static void checkValidEntityName(String entityName, int maxEntityNameLength, boolean allowSeparator) {
         if (entityName == null || entityName.isEmpty()) {
-            throw new IllegalArgumentException("Entity name cannot be null");
+            throw new IllegalArgumentException("Entity name cannot be null.");
```
Comment (conniey) [inline sdk/servicebus/microsoft-azure-servicebus/src/main/java/com/microsoft/azure/servicebus/management/EntityNameHelper.java]: Swap to Objects.requireNotNull.

File: sdk/servicebus/microsoft-azure-servicebus/src/main/java/com/microsoft/azure/servicebus/management/ManagementClientAsync.java [hunk 10]
```diff
@@ -540,7 +540,7 @@ public ManagementClientAsync(URI namespaceEndpointURI, ClientSettings clientSett
 
     private CompletableFuture<QueueDescription> putQueueAsync(QueueDescription queueDescription, boolean isUpdate) {
         if (queueDescription == null) {
-            throw new IllegalArgumentException("queueDescription passed cannot be null");
+            throw new IllegalArgumentException("queueDescription passed cannot be null.");
```
Comment (conniey) [inline sdk/servicebus/microsoft-azure-servicebus/src/main/java/com/microsoft/azure/servicebus/management/ManagementClientAsync.java]: Replace with Objects.requireNotNull

File: sdk/servicebus/microsoft-azure-servicebus/src/main/java/com/microsoft/azure/servicebus/management/QueueDescription.java [hunk 11]
```diff
@@ -73,7 +73,7 @@ public Duration getLockDuration() {
      */
     public void setLockDuration(Duration lockDuration) {
         if (lockDuration == null) {
-            throw new IllegalArgumentException("Value cannot be null");
+            throw new IllegalArgumentException("Value cannot be null.");
```
Comment (conniey) [inline sdk/servicebus/microsoft-azure-servicebus/src/main/java/com/microsoft/azure/servicebus/management/QueueDescription.java]: Replace with Objects.requireNotNull

File: sdk/servicebus/microsoft-azure-servicebus/src/main/java/com/microsoft/azure/servicebus/management/QueueDescription.java [hunk 12]
```diff
@@ -387,7 +387,7 @@ public String getUserMetadata() {
      */
     public void setUserMetadata(String userMetadata) {
         if (userMetadata == null) {
-            throw new IllegalArgumentException("Value cannot be null");
+            throw new IllegalArgumentException("Value cannot be null.");
```
Comment (conniey) [inline sdk/servicebus/microsoft-azure-servicebus/src/main/java/com/microsoft/azure/servicebus/management/QueueDescription.java]: Replace with Objects.requireNotNull

File: sdk/servicebus/microsoft-azure-servicebus/src/main/java/com/microsoft/azure/servicebus/management/SharedAccessAuthorizationRule.java [hunk 13]
```diff
@@ -55,7 +55,7 @@ public String getKeyName() {
     @Override
     public void setKeyName(String keyName) {
         if (keyName == null || keyName.isEmpty()) {
```
Comment (conniey) [inline sdk/servicebus/microsoft-azure-servicebus/src/main/java/com/microsoft/azure/servicebus/management/SharedAccessAuthorizationRule.java]: I would replace this with Replace with Objects.requireNotNull and then another check for isEmpty.

File: sdk/storage/azure-storage-common/src/main/java/com/azure/storage/common/BaseClientBuilder.java [hunk 14]
```diff
@@ -268,7 +269,7 @@ public final T httpClient(HttpClient httpClient) {
      * @throws NullPointerException If {@code pipelinePolicy} is {@code null}
      */
     public final T addPolicy(HttpPipelinePolicy pipelinePolicy) {
-        this.additionalPolicies.add(Objects.requireNonNull(pipelinePolicy));
+        this.additionalPolicies.add(Objects.requireNonNull(pipelinePolicy, "'pipelinePolicy' cannot be null"));
```
Comment (conniey) [inline sdk/storage/azure-storage-common/src/main/java/com/azure/storage/common/BaseClientBuilder.java]:
```suggestion
        this.additionalPolicies.add(Objects.requireNonNull(pipelinePolicy, "'pipelinePolicy' cannot be null."));
```

File: sdk/storage/microsoft-azure-storage-blob/src/main/java/com/microsoft/azure/storage/blob/PageBlobURL.java [hunk 15]
```diff
@@ -517,7 +517,7 @@ public PageBlobURL withSnapshot(String snapshot) throws MalformedURLException, U
         context = context == null ? Context.NONE : context;
 
         if (prevSnapshot == null) {
-            throw new IllegalArgumentException("prevSnapshot cannot be null");
+            throw new IllegalArgumentException("prevSnapshot cannot be null.");
```
Comment (conniey) [inline sdk/storage/microsoft-azure-storage-blob/src/main/java/com/microsoft/azure/storage/blob/PageBlobURL.java]: Replace with Objects.requireNotNull

**Comment by conniey** [comment]
@alexbaeza Thanks so much for contriubuting!

**Comment by alexbaeza** [comment]
Thank you @conniey for reviewing, I have applied the suggested changes as part of the next commit.

File: sdk/eventhubs/microsoft-azure-eventhubs-eph/src/main/java/com/microsoft/azure/eventprocessorhost/BaseLease.java [hunk 16]
```diff
@@ -36,22 +36,22 @@ protected BaseLease() {
      * @param partitionId Partition id for this lease.
      */
     public BaseLease(String partitionId) {
-        Objects.requireNonNull(partitionId);
+        Objects.requireNonNull(partitionId, "'patitionId' cannot be null.");
```
Comment (conniey) [inline sdk/eventhubs/microsoft-azure-eventhubs-eph/src/main/java/com/microsoft/azure/eventprocessorhost/BaseLease.java]:
```suggestion
        Objects.requireNonNull(partitionId, "'partitionId' cannot be null.");
```

File: sdk/storage/microsoft-azure-storage-blob/src/main/java/com/microsoft/azure/storage/blob/PageBlobURL.java [hunk 17]
```diff
@@ -516,9 +517,7 @@ public PageBlobURL withSnapshot(String snapshot) throws MalformedURLException, U
         accessConditions = accessConditions == null ? new BlobAccessConditions() : accessConditions;
         context = context == null ? Context.NONE : context;
 
-        if (prevSnapshot == null) {
-            throw new IllegalArgumentException("prevSnapshot cannot be null");
-        }
+        Objects.requireNonNull(prevSnapshot, "'prevSnapshot' cannot be null.")
```
Comment (conniey) [inline sdk/storage/microsoft-azure-storage-blob/src/main/java/com/microsoft/azure/storage/blob/PageBlobURL.java]: The packages that start with `microsoft-azure-*` are owned by the service team rather than us. We own the ones that are start with `azure-*`. Consequently, we try not to change existing behaviour of their client libraries. It may throw off customers if they were expecting an IllegalArgumentException but now it throws a NullPointerException.

Could you please revert this?

I just noticed the difference in packages, apologies.

File: sdk/eventgrid/microsoft-azure-eventgrid/src/main/java/com/microsoft/azure/eventgrid/customization/EventGridSubscriber.java [hunk 18]
```diff
@@ -51,12 +52,13 @@ public EventGridSubscriber() {
      */
     @Beta
     public void putCustomEventMapping(final String eventType, final Type eventDataType) {
-        if (eventType == null || eventType.isEmpty()) {
-            throw new IllegalArgumentException("eventType parameter is required and cannot be null or empty");
-        }
-        if (eventDataType == null) {
-            throw new IllegalArgumentException("eventDataType parameter is required and cannot be null");
+        Objects.requireNonNull(eventType, "'eventType' cannot be null.");
```
Comment (conniey) [inline sdk/eventgrid/microsoft-azure-eventgrid/src/main/java/com/microsoft/azure/eventgrid/customization/EventGridSubscriber.java]: I just noticed the package name this change exists in.  The packages that start with `microsoft-azure-*` are owned by the service team rather than us. We own the ones that are start with `azure-*` (ie. azure-core, azure-messaging-eventhubs). Consequently, we try not to change existing behaviour of their client libraries.

Could you please revert this?

File: sdk/eventhubs/microsoft-azure-eventhubs/src/main/java/com/microsoft/azure/eventhubs/RetryPolicy.java [hunk 19]
```diff
@@ -8,6 +8,7 @@
```
Comment (conniey) [inline sdk/eventhubs/microsoft-azure-eventhubs/src/main/java/com/microsoft/azure/eventhubs/RetryPolicy.java]: Could you please revert this change in behaviour?

File: sdk/eventhubs/microsoft-azure-eventhubs/src/main/java/com/microsoft/azure/eventhubs/impl/EventDataBatchImpl.java [hunk 20]
```diff
@@ -12,6 +12,7 @@
 import java.util.LinkedList;
```
Comment (conniey) [inline sdk/eventhubs/microsoft-azure-eventhubs/src/main/java/com/microsoft/azure/eventhubs/impl/EventDataBatchImpl.java]: Could you please revert this change in behaviour?

File: sdk/eventhubs/microsoft-azure-eventhubs/src/main/java/com/microsoft/azure/eventhubs/impl/EventDataImpl.java [hunk 21]
```diff
@@ -39,9 +39,8 @@ private EventDataImpl() {
```
Comment (conniey) [inline sdk/eventhubs/microsoft-azure-eventhubs/src/main/java/com/microsoft/azure/eventhubs/impl/EventDataImpl.java]: Could you please revert this change in behaviour?

File: sdk/servicebus/microsoft-azure-servicebus/src/main/java/com/microsoft/azure/servicebus/management/SharedAccessAuthorizationRule.java [hunk 22]
```diff
@@ -54,8 +55,10 @@ public String getKeyName() {
 
     @Override
     public void setKeyName(String keyName) {
-        if (keyName == null || keyName.isEmpty()) {
-            throw new IllegalArgumentException("Argument cannot be null");
+        Objects.requireNonNull(keyName, "'keyName' cannot be null.");
```
Comment (conniey) [inline sdk/servicebus/microsoft-azure-servicebus/src/main/java/com/microsoft/azure/servicebus/management/SharedAccessAuthorizationRule.java]: Could you please revert this change in behaviour?

File: sdk/servicebus/microsoft-azure-servicebus/src/main/java/com/microsoft/azure/servicebus/management/ManagementClientAsync.java [hunk 23]
```diff
@@ -539,9 +540,7 @@ public ManagementClientAsync(URI namespaceEndpointURI, ClientSettings clientSett
     }
 
     private CompletableFuture<QueueDescription> putQueueAsync(QueueDescription queueDescription, boolean isUpdate) {
-        if (queueDescription == null) {
-            throw new IllegalArgumentException("queueDescription passed cannot be null");
-        }
+        Objects.requireNonNull(queueDescription, "'queueDescription' cannot be null.");
```
Comment (conniey) [inline sdk/servicebus/microsoft-azure-servicebus/src/main/java/com/microsoft/azure/servicebus/management/ManagementClientAsync.java]: Could you please revert this change in behaviour?

File: sdk/servicebus/microsoft-azure-servicebus/src/main/java/com/microsoft/azure/servicebus/management/EntityNameHelper.java [hunk 24]
```diff
@@ -81,8 +83,10 @@ static void checkValidRuleName(String ruleName) {
     }
 
     private static void checkValidEntityName(String entityName, int maxEntityNameLength, boolean allowSeparator) {
-        if (entityName == null || entityName.isEmpty()) {
-            throw new IllegalArgumentException("Entity name cannot be null");
+        Objects.requireNonNull(entityName, "'entityName' cannot be null.");
```
Comment (conniey) [inline sdk/servicebus/microsoft-azure-servicebus/src/main/java/com/microsoft/azure/servicebus/management/EntityNameHelper.java]: Could you please revert this change in behaviour?

**Comment by alexbaeza** [comment]
@conniey Thanks again for the feedback I have reverted the changes as suggested :)

File: sdk/core/azure-core/src/main/java/com/azure/core/util/Context.java [hunk 25]
```diff
@@ -39,11 +40,8 @@
      * @throws IllegalArgumentException If {@code key} is {@code null}.
      */
     public Context(Object key, Object value) {
-        if (key == null) {
-            throw new IllegalArgumentException("key cannot be null");
-        }
         this.parent = null;
-        this.key = key;
+        this.key = Objects.requireNonNull(key, "'key' cannot be null");
```
Comment (hemanttanwar) [inline sdk/core/azure-core/src/main/java/com/azure/core/util/Context.java]: nitPick: Like others, can we add fullstop in the end  i.e. "'key' cannot be null."

Comment (alexbaeza) [reply sdk/core/azure-core/src/main/java/com/azure/core/util/Context.java]: of course :)

File: sdk/core/azure-core/src/main/java/com/azure/core/http/rest/PagedFluxBase.java [hunk 26]
```diff
@@ -68,8 +68,8 @@ public PagedFluxBase(Supplier<Mono<P>> firstPageRetriever) {
      */
     public PagedFluxBase(Supplier<Mono<P>> firstPageRetriever,
                          Function<String, Mono<P>> nextPageRetriever) {
-        Objects.requireNonNull(firstPageRetriever, "First page supplier cannot be null");
-        Objects.requireNonNull(nextPageRetriever, "Next page retriever function cannot be null");
+        Objects.requireNonNull(firstPageRetriever, "First page supplier cannot be null.");
```
Comment (srnagar) [inline sdk/core/azure-core/src/main/java/com/azure/core/http/rest/PagedFluxBase.java]: nit: This can be made consistent with the rest to use the variable name instead of the description.
```suggestion
        Objects.requireNonNull(firstPageRetriever, "'firstPageRetriever' cannot be null.");
```

File: sdk/core/azure-core/src/main/java/com/azure/core/http/rest/PagedFluxBase.java [hunk 27]
```diff
@@ -68,8 +68,8 @@ public PagedFluxBase(Supplier<Mono<P>> firstPageRetriever) {
...
+        Objects.requireNonNull(nextPageRetriever, "Next page retriever function cannot be null.");
```
Comment (srnagar) [inline sdk/core/azure-core/src/main/java/com/azure/core/http/rest/PagedFluxBase.java]:
```suggestion
        Objects.requireNonNull(nextPageRetriever, "'nextPageRetriever' function cannot be null.");
```

File: sdk/keyvault/azure-keyvault-certificates/src/main/java/com/azure/security/keyvault/certificates/CertificateClient.java [hunk 28]
```diff
@@ -1256,7 +1256,7 @@ public Certificate mergeCertificate(MergeCertificateOptions mergeCertificateConf
      */
     @ServiceMethod(returns = ReturnType.SINGLE)
     public Response<Certificate> mergeCertificateWithResponse(MergeCertificateOptions mergeCertificateConfig, Context context) {
-        Objects.requireNonNull(mergeCertificateConfig, "The merge certificate configuration cannot be null");
+        Objects.requireNonNull(mergeCertificateConfig, "The merge certificate configuration cannot be null.");
```
Comment (srnagar) [inline sdk/keyvault/azure-keyvault-certificates/src/main/java/com/azure/security/keyvault/certificates/CertificateClient.java]:
```suggestion
        Objects.requireNonNull(mergeCertificateConfig, "'mergeCertificateConfig' cannot be null.");
```

File: sdk/keyvault/azure-keyvault-certificates/src/main/java/com/azure/security/keyvault/certificates/CertificateAsyncClient.java [hunk 29]
```diff
@@ -935,7 +935,7 @@ private CertificatePolicy getDefaultPolicy() {
      */
     @ServiceMethod(returns = ReturnType.SINGLE)
     public Mono<Response<Certificate>> mergeCertificateWithResponse(MergeCertificateOptions mergeCertificateConfig) {
-        Objects.requireNonNull(mergeCertificateConfig, "The merge certificate configuration cannot be null");
+        Objects.requireNonNull(mergeCertificateConfig, "The merge certificate configuration cannot be null.");
```
Comment (srnagar) [inline sdk/keyvault/azure-keyvault-certificates/src/main/java/com/azure/security/keyvault/certificates/CertificateAsyncClient.java]:
```suggestion
        Objects.requireNonNull(mergeCertificateConfig, "'mergeCertificateConfig' cannot be null.");
```

File: sdk/keyvault/azure-keyvault-certificates/src/main/java/com/azure/security/keyvault/certificates/CertificateClientBuilder.java [hunk 30]
```diff
@@ -164,7 +164,7 @@ public CertificateClientBuilder endpoint(String endpoint) {
      * @throws NullPointerException if {@code credential} is {@code null}.
      */
     public CertificateClientBuilder credential(TokenCredential credential) {
-        Objects.requireNonNull(credential, "Token credential cannot be null");
+        Objects.requireNonNull(credential, "Token credential cannot be null.");
```
Comment (srnagar) [inline sdk/keyvault/azure-keyvault-certificates/src/main/java/com/azure/security/keyvault/certificates/CertificateClientBuilder.java]:
```suggestion
        Objects.requireNonNull(credential, "'credential' cannot be null.");
```

File: sdk/keyvault/azure-keyvault-certificates/src/main/java/com/azure/security/keyvault/certificates/CertificateClientBuilder.java [hunk 31]
```diff
@@ -190,7 +190,7 @@ public CertificateClientBuilder httpLogOptions(HttpLogOptions logOptions) {
      * @throws NullPointerException if {@code policy} is {@code null}.
      */
     public CertificateClientBuilder addPolicy(HttpPipelinePolicy policy) {
-        Objects.requireNonNull(policy, "Http pipeline policy cannot be null");
+        Objects.requireNonNull(policy, "Http pipeline policy cannot be null.");
```
Comment (srnagar) [inline sdk/keyvault/azure-keyvault-certificates/src/main/java/com/azure/security/keyvault/certificates/CertificateClientBuilder.java]:
```suggestion
        Objects.requireNonNull(policy, "'policy' cannot be null.");
```

File: sdk/core/azure-core/src/main/java/com/azure/core/http/rest/PagedFluxBase.java (hunk 26 above)
Comment (alexbaeza) [reply sdk/core/azure-core/src/main/java/com/azure/core/http/rest/PagedFluxBase.java]: Sure, Thank you fixing it now :)
